cumulusci/oauth/client.py

```diff
@@ -0,0 +1,293 @@
+from pathlib import Path
+from urllib.parse import parse_qs
+from urllib.parse import quote
+from urllib.parse import urlparse
+
+from cumulusci.core.exceptions import CumulusCIUsageError
+from cumulusci.oauth.client_info import OAuthClientInfo
+from cumulusci.oauth.exceptions import OAuthError
+from cumulusci.utils.http.requests_utils import safe_json_from_response
+
+logger = logging.getLogger(__name__)
+
+HTTP_HEADERS = {"Content-Type": "application/x-www-form-urlencoded"}
+SANDBOX_DOMAIN_RE = re.compile(
+    r"^https://([\w\d-]+\.)?(test|cs\d+)(\.my)?\.salesforce\.com/?$"
+)
+SANDBOX_LOGIN_URL = (
+    os.environ.get("SF_SANDBOX_LOGIN_URL") or "https://test.salesforce.com"
+)
+PROD_LOGIN_URL = os.environ.get("SF_PROD_LOGIN_URL") or "https://login.salesforce.com"
+
+
+def create_pem_files():
+    """Create both a localhost.pem and key.pem file"""
+    key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
+    subject = issuer = x509.Name(
+        [
+            x509.NameAttribute(NameOID.COUNTRY_NAME, "US"),
+            x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, "California"),
+            x509.NameAttribute(NameOID.LOCALITY_NAME, "San Francisco"),
+            x509.NameAttribute(NameOID.ORGANIZATION_NAME, "Salesforce.com"),
+        ]
+    )
+    cert = (
+        x509.CertificateBuilder()
+        .subject_name(subject)
+        .issuer_name(issuer)
+        .public_key(key.public_key())
+        .serial_number(x509.random_serial_number())
+        .not_valid_before(datetime.now())
+        .not_valid_after(datetime.now() + timedelta(hours=1))
+        .sign(key, hashes.SHA256())
+    )
+    with open("localhost.pem", "w") as f:
+        f.write(cert.public_bytes(serialization.Encoding.PEM).decode("latin1"))
+
+    with open("key.pem", "w") as f:
+        f.write(
+            key.private_bytes(
+                serialization.Encoding.PEM,
+                serialization.PrivateFormat.TraditionalOpenSSL,
+                serialization.NoEncryption(),
+            ).decode("latin1")
+        )
+
+
+class OAuth2Client(object):
+    """Represents an OAuth2 client with the ability to
+    execute different grant types.
+    Currenntly supported grant types include:
+        (1) Authorization Code - via auth_code_flow()
+
+    To instantiate, just provide an instance or subclass
+    of OAuthClientInfo to the contructor.
+    """
+
+    def __init__(self, client_info: OAuthClientInfo):
+        self.client_info = client_info
+        self.response = None
+        self.httpd = None
+        self.httpd_timeout = 300
+
+    def auth_code_flow(self, use_https=False) -> None:
+        """Completes the flow for the OAuth2 auth code grant type.
+        For more info on the auth code flow see:
+        https://www.oauth.com/oauth2-servers/server-side-apps/authorization-code/
+
+        @param auth_info instanace of OAuthInfo to use in the flow
+        @param use_https wether or not the http daemon receiving the callback
```

Comment:
Can't we determine this automatically from the callback url?

Reply:
ah, yes. This would be much nicer!